### ISSUE TITLE
Rc variant and compressing pointer size

### ISSUE THREAD

**Issue by amarmaduke** [AdrienChampion/hashconsing#13]
Hi,

I've been expermenting with hconsing with NBE for lambda calculus. This library is the primary inspiration, which I've mostly copied. Originally, I was using the library, but ran into some performance problems:

1. I believe the `uid` is not needed, the internal `NonNull` pointer data in the `Rc` is enough to act as a unique reference. As every `Rc` is the spoke on a fuzzy ball surrounding a `RcBox`, even if the location of the `RcBox` were to somehow move, it must move in a way such that all constructed `Rc`'s remain sound. Of course, the `Weak` stored in the `HConsign` has the same properties, because it references the same `RcBox`. Moreover, once the pointer is reclaimed, that is the exact condition in which the unique id would be reusable because the original `Weak` would fail to upgrade.
2. The usage of `Arc` with no ability to switch to an `Rc`. In my case, there is no point in being multithreaded. Parallelization in evaluating lambda terms is not very good unless you take an optimal lambda evaluator route.

"Fixing" those two problems were about a 2-3x speedup for my benchmarks, especially because lambda terms are explosive in terms of size, so halving the pointer size ends up being a useful win.

Ideally, I'd like these changes to make it into this library, because then I can just pull it from cargo :)

Unrelated, I would prefer if the pointer types where called `Hc` and `Ahc`, but this is a minor thing that is not important to me, I can just rename the types in my project.

If you are interested I could draft a PR with the above changes.

**Comment by Pat-Lafon** [comment]
I would be interested in seeing this as a pr, even if just to add this to my own fork. Do you have a benchmark you can add to the repository to show the speedup?

**Comment by AdrienChampion** [comment]
Thank (both) for your PR/issues ❤️ 
I will finally have time to spend on hashconsing this week and go over your feedback and contributions

**Comment by AdrienChampion** [comment]
> The usage of Arc with no ability to switch to an Rc. In my case, there is no point in being multithreaded. Parallelization in > evaluating lambda terms is not very good unless you take an optimal lambda evaluator route.

So do you not use a static factory? If you do, what do you store in the map?

Other than that, I think @Pat-Lafon's [PR on having a derive-macro](https://github.com/AdrienChampion/hashconsing/pull/14) could help by having an attribute for that eventually. What do you think?

**Comment by AdrienChampion** [comment]
Regarding the `uid`, I see you point. Could you provide a link to your version so that I can take a look?

**Comment by amarmaduke** [comment]
Here is a gist of what I'm currently using: https://gist.github.com/amarmaduke/d5abb2cc8f7f1b769f4af661bf701a32

I'm not using a static store. I pass around a container of all application state via mutable reference, the hashcons table is stored there.

Another ask (that I'm willing to contribute): `into_raw` and `from_raw` mirroring the `Rc` variants. Worse case these leak memory by not updating the strong-count, however, it's useful for me because I need to emulate a call stack (Rust is not great with recursion) and I would like to do it in a type-agonistic way.

**Comment by AdrienChampion** [comment]
Thank you for the link, I will take a look.

> I'm not using a static store. I pass around a container of all application state via mutable reference, the hashcons table is stored there.

Yeah I'm not surprised, that's the problem with `Rc`: you can't put them in a static storage. So we could have a derive-attribute that would exclude static-factory stuff.

**Comment by AdrienChampion** [comment]
@amarmaduke I'm a bit concerned with [your hash function](https://gist.github.com/amarmaduke/d5abb2cc8f7f1b769f4af661bf701a32#file-hc-rs-L34-L36).

It seems to me like like whenever a term is moved by memory management its hash will change. It will stay equal to itself, but if its placement in the factory relies on a previous value of the pointer then you won't find it since the hash has change.

Does this make sense or am I speaking nonsense?

**Comment by AdrienChampion** [comment]
Thank you for this great answer. I'm not an expert in pinning, I'm going to research this a bit on my end to convince myself.

**Comment by AdrienChampion** [comment]
Short term, how about this: we can have a branch on this repo that's the same but without the `u64` uid.

That way you can use this branch instead of your fork, and we can run experiments on a WIP version to make sure there's no problem.

**Comment by AdrienChampion** [comment]
Alternatively, we could just feature-gate the feature in the `main` branch and say it's experimental somewhere as long as we're not sure it's safe

**Comment by AdrienChampion** [comment]
It's starting to sound like you're right:
- https://users.rust-lang.org/t/can-a-rc-move-location-behind-my-back/28828/2

Since hashconsing will not let you mutate terms anyways, I think addresses must be stable like you said. Inner mutability would potentially break this, but it breaks hashconsing altogether anyways. That should probably be documented somewhere actually `>_<`

**Comment by AdrienChampion** [comment]
By the way @amarmaduke, it sounds like we will have quite a few configuration options soon, so I think @Pat-Lafon's PR #14 is the way to go. We would have a derive macro where you could (de)activate `Arc`/`Rc`, static factory and so on. It would also probably generate the `HConsed` type directly in your crate so that you can decide how it's called and provide implementations.

Would that work for you?

**Comment by amarmaduke** [comment]
Yeah, that would likely work for me, worst case it is future-proof to more knobs :)

**Comment by AdrienChampion** [comment]
My thoughts exactly 😸 

Let me take care of the proc-macro PR first then
